Feature/populate meme to database

fnoori opened the pull request:
- Saving meme data to database now works
- arguments 
    - meme: title, cloudinary_url
    - token: auth token

Commit:
Store meme info in db now works

The socket data must have a token and The meme information (title, cloudinary_url); the user_id is taken from the token

Commit:
Renamed 'save' variable for clarity

nxtman123 reviewed (changes requested):
3 Things:
- I've added CI integration, so merge develop into your branch to get that working on this Pull Request
  - pushing new commits to your branch will update this PR automatically, so don't close it
- remember to lint
- user information should be kept in memory on the server in the socket scope so we don't have to veryify auth for everything they do. This is different from normal stateless http, because we have sockets with a persistent connection, the server should remember who this socket is connected to.
  - for a logged in and reconnecting client, it should pass the token it's stored when connecting the socket, so check `socket.handshake.query.myTokenVar` after `console.log('a user connected!');`

Commit:
Merge branch 'develop' into feature/populate_meme_to_database

Commit:
Stateful authentication

Token is checked on connect

On successful login, token is checked and state is also set

nxtman123 commented on src/server/main.js:
```diff
@@ -63,9 +70,25 @@ io.on('connect', (socket) => {
+    if (saveResult !== true) {
+      return socket.emit('uploadMemeData', saveResult);
+    }
+    return socket.emit('uploadMemeData', 'successfully saved meme');
```
On the "happy path" return something the client can more reliably test like `true`. So you could probably remove the `if` block and last line, and go with the line inside the `if`

nxtman123 commented on src/server/main.js:
```diff
@@ -63,9 +70,25 @@ io.on('connect', (socket) => {
    */
   socket.on('login', async (user) => {
-    const login = await authentication.login(user);
+    const token = await authentication.login(user);
```
this `token` is what should be tested. It could be a `String` if there was an error

fnoori replied:
I implemented the changes in the comment immediately below this one

In this case, even if the token is an error ```String```, when calling 
```socketUser = authentication.verifyToken(token);```, this verification will return a ```false```.

The output of ```socketUser = authentication.verifyToken(token);``` is checked, if it's ```false```, then it will return an error to the user, otherwise it will return the token

nxtman123 commented on src/server/main.js:
```diff
@@ -63,9 +70,25 @@ io.on('connect', (socket) => {
   socket.on('login', async (user) => {
-    const login = await authentication.login(user);
+    const token = await authentication.login(user);
+    const tokenVerification = authentication.verifyToken(token);
```
```suggestion
    socketUser = authentication.verifyToken(token);
```

nxtman123 commented on src/server/main.js:
```diff
@@ -63,9 +70,25 @@ io.on('connect', (socket) => {
+
+    if (tokenVerification !== false) {
+      socketUser = authentication.verifyToken(socket.handshake.query.token);
+    }
```
remove this block

Commit:
Implemented recommendations

Clarified the variable names

Removed unnecisary if-else

nxtman123 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Populate meme information to meme table in database](https://trello.com/c/fGA1CJM0/28-populate-meme-information-to-meme-table-in-database)

Commit:
variable name clarification

adnanrh commented on src/server/main.js:
```diff
@@ -63,9 +70,22 @@ io.on('connect', (socket) => {
+    if (socketUser === false) {
+      return socket.emit('uploadMemeData', 'cannot verify user');
+    }
+    const saveResult = await meme.saveMeme(data, socketUser);
```
This event might need refactoring if we decide to send the image through socket io. If we upload to cloudinary from the client instead, then this won't need any changes.

fnoori replied:
That is true, but I think uploading the file to cloudinary would be easier if done from the frontend.

Commit:
Further removal of unnecessary code
